File: java/src/main/java/com/urbanlogiq/ulsdk/TestContext.java
// Copyright (c), CommunityLogiq Software

package com.urbanlogiq.ulsdk;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.HashMap;
import java.util.List;
import com.urbanlogiq.ulsdk.ApiKeyContext;

public final class TestContext extends RequestContext {
    ApiKeyContext _context;
    byte[] _response;

    public Region region() {
        return this._context.region();
    }

    public Environment environment() {
        return this._context.environment();
    }

    public TestContext(ApiKeyContext context) {
        this._context = context;
    }

    public void setResponse(byte[] response) {
        this._response = response;
    }

    public byte[] get(String path, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String echoPath = "/v1/echo/";
        this._context.get(echoPath, params, headers);
        byte[] response = this._response;
        this._response = null;
        return response;
    }

    public byte[] put(String path, byte[] body, String mimetype, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String echoPath = "/v1/echo/";
        
        // roundtrip the _expected result_ through the echo endpoint to make
        // sure that a) we get it back unchanged and b) we can deserialize it
        // properly.
        byte[] response = this._context.put(echoPath, this._response, mimetype, params, headers);

        if (this._response != null) {
            if (!java.util.Arrays.equals(response, this._response)) {
                throw new RuntimeException("Test failure, expected response to match request");
            }
        }

        this._response = null;
        return response;
    }

    public byte[] post(String path, byte[] body, String mimetype, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String echoPath = "/v1/echo/";

        // roundtrip the _expected result_ through the echo endpoint to make
        // sure that a) we get it back unchanged and b) we can deserialize it
        // properly.
        byte[] response = this._context.post(echoPath, this._response, mimetype, params, headers);

        if (this._response != null) {
            if (!java.util.Arrays.equals(response, this._response)) {
                throw new RuntimeException("Test failure, expected response to match request");
            }
        }

        this._response = null;
        return response;
    }

    public byte[] upload(String path, List<File> files) throws URISyntaxException, IOException, InterruptedException {
        String echoPath = "/v1/echo/";
        this._context.upload(echoPath, files);
        byte[] response = this._response;
        this._response = null;
        return response;
    }

    public byte[] delete(String path, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String echoPath = "/v1/echo/";
        this._context.delete(echoPath, params, headers);

        byte[] response = this._response;
        this._response = null;
        return response;
    }
}

File: rust/ulsdk/src/request_context/test_context.rs
// Copyright (c), CommunityLogiq Software

use async_trait::async_trait;
use bytes::Bytes;

use crate::error::Error;
use crate::request_context::{ApiKeyContext, File, HeaderMap, ParamMap, RequestContext};
use crate::{Environment, Region};

pub(crate) struct TestContext {
    context: ApiKeyContext,
    response: Vec<u8>,
}

impl TestContext {
    pub(crate) fn new(context: ApiKeyContext) -> Self {
        Self {
            context,
            response: Vec::new(),
        }
    }

    pub(crate) fn set_response<T>(&mut self, r: T)
    where
        T: Into<Vec<u8>>,
    {
        self.response = r.into();
    }
}

#[async_trait]
impl RequestContext for TestContext {
    fn region(&self) -> Region {
        self.context.region()
    }

    fn environment(&self) -> Environment {
        self.context.environment()
    }

    async fn get(
        &self,
        _path: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let echo_path = "/v1/echo/";

        self.context.get(echo_path, params, headers).await?;
        Ok(self.response.clone())
    }

    async fn put(
        &self,
        _path: &str,
        _body: Bytes,
        mimetype: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let echo_path = "/v1/echo/";

        let response = self.response.clone();
        let response_bytes = Bytes::from(response.clone());
        let r = self
            .context
            .put(echo_path, response_bytes, mimetype, params, headers)
            .await?;

        if r != response {
            return Err(Error::Unclassified(
                "Test failure, expected response to match request".into(),
            ));
        }

        Ok(response)
    }

    async fn post(
        &self,
        _path: &str,
        _body: Bytes,
        mimetype: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let echo_path = "/v1/echo/";

        let response = self.response.clone();
        let response_bytes = Bytes::from(response.clone());
        let r = self
            .context
            .post(echo_path, response_bytes, mimetype, params, headers)
            .await?;

        if r != response {
            return Err(Error::Unclassified(
                "Test failure, expected response to match request".into(),
            ));
        }

        Ok(response)
    }

    async fn upload(&self, _path: &str, files: Vec<File>) -> Result<Vec<u8>, Error> {
        let echo_path = "/v1/echo/";

        self.context.upload(echo_path, files).await?;

        Ok(self.response.clone())
    }

    async fn delete(
        &self,
        _path: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let echo_path = "/v1/echo/";

        self.context.delete(echo_path, params, headers).await?;
        Ok(self.response.clone())
    }
}
